LG-4637: Show accurate time remaining for IAL2 throttle screens

aduth opened the pull request:
**Why**: As a user, I expect that if I encounter an error screen informing me that I am unable to proceed due to too many attempts, the text on the page will let me know when I can try again, so that I know when to return to try again, and so that I'm not waiting longer than necessary if I can try again sooner than the text would otherwise imply.

**Screenshot:**

Screen|Before|After
---|---|---
Session Error|![Screen Shot 2021-08-17 at 3 17 37 PM](https://user-images.githubusercontent.com/1779930/129787520-3a570a20-6995-4bfe-95ac-52f5b8fa1f44.png)|![Screen Shot 2021-08-17 at 3 17 25 PM](https://user-images.githubusercontent.com/1779930/129787519-60a62b57-6d29-40e3-ac97-3d22121bf61d.png)
SSN Error|![Screen Shot 2021-08-17 at 3 17 37 PM](https://user-images.githubusercontent.com/1779930/129787520-3a570a20-6995-4bfe-95ac-52f5b8fa1f44.png)|![Screen Shot 2021-08-31 at 8 41 40 AM](https://user-images.githubusercontent.com/1779930/131507607-92449573-3e2a-4ace-9fee-ed62845b65df.png)
Phone Error|![Screen Shot 2021-08-31 at 8 59 12 AM](https://user-images.githubusercontent.com/1779930/131507634-ba31b40b-f146-47d2-a0fb-984f737ec789.png)|![Screen Shot 2021-08-31 at 9 00 07 AM](https://user-images.githubusercontent.com/1779930/131507647-ccb9a835-a899-4b68-a3ae-328782ae4cf3.png)
Send Link|![localhost_3000_verify_doc_auth_send_link](https://user-images.githubusercontent.com/1779930/130249003-2b8a04d5-41f4-4876-8f31-940d8e2dca8b.png)|![localhost_3000_verify_doc_auth_send_link (1)](https://user-images.githubusercontent.com/1779930/130249016-3c20bbca-e8bf-40fc-9a64-53babda365a3.png)
GPO Verify|![localhost_3000_account_verify](https://user-images.githubusercontent.com/1779930/130249038-d184d487-9899-4d56-952d-27a9ed04110a.png)|![localhost_3000_account_verify (2)](https://user-images.githubusercontent.com/1779930/130249456-84fbddc2-b5b6-4486-8d5d-72a691bac83c.png)

aduth commented on app/models/throttle.rb:
```diff
@@ -54,23 +54,35 @@ def throttled?
-    config = THROTTLE_CONFIG.with_indifferent_access[throttle_type]
-    [config[:max_attempts], config[:attempt_window]]
+  def self.expires_at(conditions)
+    Enumerator.new do |e|
+      e << find_by(conditions)&.expires_at
+      e << attempt_window_in_minutes(conditions[:throttle_type]) if conditions[:throttle_type]
+      e << 0
+    end.find(&:present?)
```
I might be pushing into bad-clever territory here. The general thinking was there's a few different places we can try to get the expiration from, and we'd want to evaluate them lazily.

I suppose a plainer approach is mostly similar now that I think of it...

```suggestion
    record_expires_at = find_by(conditions)&.expires_at
    return record_expires_at if record_expires_at.present?
    return attempt_window_in_minutes(conditions[:throttle_type]) if conditions[:throttle_type]
    0
```

aduth replied:
Ah, I think the fallback ought to be `Time.zone.now` ? Originally this was implemented as a `throttled_time_remaining` method, returning the minutes 'til the throttle expires.

zachmargolis replied:
I'm a fan of the expanded if statements, but I also have one more idea for lazy evaluation:

```suggestion
    find_by(conditions)&.expires_at.presence ||
      (conditions[:throttle_type] && attempt_window_in_minutes(conditions[:throttle_type])).presence ||
      0
```

aduth replied:
In the rebased bd57718, I dropped the class method and just call it from the throttle instance via `Throttle.for`.

zachmargolis commented on config/locales/dotiw/en.yml:
```diff
@@ -2,4 +2,4 @@
   datetime:
     dotiw:
-      two_words_connector: and
+      two_words_connector: ' and '
```
I feel like we go back and forth on the spaces in this one all the time...?

aduth replied:
> I feel like we go back and forth on the spaces in this one all the time...?

Heh, I had a feeling I'd seen this before as well. Let me see if I can dig up any relevant history.

aduth replied:
Ah, I think we've seen it recently because we tend to overwrite this from the call to `distance_of_time_in_words`.

For example:

https://github.com/18F/identity-idp/blob/5d12301efa96f715ec0a1e6fb763502a633ee521/app/helpers/session_timeout_warning_helper.rb#L21-L27

We could do the same here, although it kinda feels like it might be better to tackle it at the source? We'd want to update these other overrides if we do, otherwise we'll get some "3 hours&nbsp;&nbsp;and&nbsp;&nbsp;2 minutes" excessive spacing issues.

aduth replied:
For reference, the default is " and ", with the spaces:

https://github.com/radar/distance_of_time_in_words#two_words_connector

I don't know that we had any strong reason to omit the spaces.

zachmargolis replied:
I think that part of the issue was a fight of custom strings plus our limited YAML normalization at the time (removing trailing spaces, etc), but I agree seems better to update at the source

aduth replied:
Yep, I think that's exactly right (https://github.com/18F/identity-idp/pull/3780#discussion_r431246132). I'd like to think our fancy new YAML normalizer can handle this now. It's worth a shot!

aduth replied:
This was addressed separately in #5312 and dropped in the rebased bd57718.

zachmargolis commented on app/models/throttle.rb:
```diff
@@ -54,23 +54,35 @@ def throttled?
+  end
+
+  def self.attempt_window_in_minutes(throttle_type)
+    THROTTLE_CONFIG.with_indifferent_access.dig(throttle_type, :attempt_window)
```
WDYT of just making the `THROTTLE_CONFIG` hash have indifferent access?

```
THROTTLE_CONFIG = {
  # ... etc
}.with_indifferent_access.freeze
```

aduth replied:
> WDYT of just making the `THROTTLE_CONFIG` hash have indifferent access?

I like it! Originally I tried removing the `with_indifferent_access` altogether, but quickly discovered it broke a lot of existing code.

aduth replied:
> WDYT of just making the `THROTTLE_CONFIG` hash have indifferent access?

Updated in the rebased bd57718.

zachmargolis commented on spec/models/throttle_spec.rb:
```diff
@@ -135,10 +166,42 @@
   end
 
+  describe '#expires_at' do
+    before { Timecop.freeze }
+    after { Timecop.return }
```
I like the around filter version of timecop so we don't have to worry about matching a before with an after

```suggestion
    around do |ex|
      Timecop.freeze { ex.run }
    end
```

aduth replied:
>I like the around filter version of timecop 

Ooh, I like it!

aduth commented:
There's another instance of this message I still need to deal with for phone failure:

https://github.com/18F/identity-idp/blob/439d2700911df6250955f8ab72385e4c6c8c0bab/app/views/idv/phone_errors/failure.html.erb#L24-L26

This looks a bit complicated to address, because of how there seems to be a separate throttling mechanism using the IDV session for managing this ([example](https://github.com/18F/identity-idp/blob/439d2700911df6250955f8ab72385e4c6c8c0bab/app/controllers/concerns/idv_step_concern.rb#L18-L20)). I'm not sure if there was a compelling reason to do it this way, or if we could migrate these to using the standard throttling behaviors.

zachmargolis reviewed (approved):
LGTM

aduth commented on app/controllers/idv/session_errors_controller.rb:
```diff
@@ -7,16 +7,24 @@ class SessionErrorsController < ApplicationController
+      @expires_at = throttle.expires_at
+    end
+
+    def throttled
+      @expires_at = throttle.expires_at
```
`throttled` is for the `:idv_acuant` throttle type, so this isn't quite correct. Will fix in revisions accounting for separate SSN throttle via #5346. I'll also double-check to see why feature tests didn't flag this.

aduth replied:
> I'll also double-check to see why feature tests didn't flag this.

There were feature specs for this, but they weren't specifically checking the content as showing an expected time remaining. I've added an extra set of assertions in the rebased 1612180 ([see `document_capture_step_spec.rb`](https://github.com/18F/identity-idp/pull/5301/commits/1612180d84337a1a5995182f328dc4f597f477ac#diff-07a84ad3b4f5779028c66ce93ded1dd1ddb0644ae540571c112c242bf9ac95d4)).

aduth commented:
I rebased this to incorporate revisions from #5346 and #5339, and updated original comment to include before / after screenshots for the phone and SSN throttling error screens.

aduth commented on app/controllers/idv/session_errors_controller.rb:
```diff
@@ -7,20 +7,35 @@ class SessionErrorsController < ApplicationController
-        user: effective_user,
-        throttle_type: :idv_resolution,
-      ).remaining_count
+    def ssn_from_doc
+      user_session&.dig('idv/doc_auth', 'pii_from_doc', 'ssn')
```
A bit more involved to get at the throttle target value for SSN throttling, vs. others just getting by the current user.
